test: fix possible goroutine leaks in unit tests

lzhfromustc opened the pull request:
***Description***
This pull request fixes a few goroutine leak problems when unit tests fail. 
When `t.Fatal()` is called, only the goroutine calling it will be killed, which may leave other goroutines in these three unit tests blocking and occupying additional resources.

***The fix in internal/resolver/config_selector_test.go***
This unit test contains four channels that are used in another child goroutine running `cs1` or `cs2`.
If the unit test calls `t.Fatal()`, the child goroutine will be blocked at an operation of one of the four channels.

Since the child goroutine only uses receive of `retChan1` and `retChan2`, we defer the close of them in main goroutine.
Since the child goroutine only uses send of `cs1Called` and `cs2Called` for one time, we add 1 buffer to them.

***The fix in xds/internal/httpfilter/fault/fault_test.go***
The channel `streams` has a send at the end of , and a receive in `endStream()`.
However, when one of the `endStream()` calls `t.Fatal()`, one of the `startStream()` may block.

Since `startStream()` is called 5 times, we add 5 buffer to the channel `streams`.

***The fix in xds/internal/server/listener_wrapper_test.go***
The channel `lis.acceptCh` is used to pass `connAndErr`.
However, when the main testing goroutine timeouts, the goroutine below will be blocked at `cne := <-fl.acceptCh` on line 120:
https://github.com/grpc/grpc-go/blob/83f9def5feb388c4fd7e6586bd55cf6bf6d46a01/xds/internal/server/listener_wrapper_test.go#L283-L284
https://github.com/grpc/grpc-go/blob/83f9def5feb388c4fd7e6586bd55cf6bf6d46a01/xds/internal/server/listener_wrapper.go#L192-L203
https://github.com/grpc/grpc-go/blob/83f9def5feb388c4fd7e6586bd55cf6bf6d46a01/xds/internal/server/listener_wrapper_test.go#L119-L122

To fix this problem, add a `defer close(lis.acceptCh)`. Now when the main goroutine timeouts, the child goroutine will return a non-temporary error and safely return.

RELEASE NOTES: none

dfawley commented on xds/internal/httpfilter/fault/fault_test.go:
```diff
@@ -608,7 +608,7 @@ func (s) TestFaultInjection_MaxActiveFaults(t *testing.T) {
 	defer cancel()
 
-	streams := make(chan testpb.TestService_FullDuplexCallClient)
+	streams := make(chan testpb.TestService_FullDuplexCallClient, 5) // startStream() is called 5 times
```
For this one, let's just change the `Fatalf` to an `Errorf`; I don't think it's a problem to continue running the rest of the test if an error happens.

dfawley reviewed (changes requested):
Thanks for the fixes!  I have one minor suggestion below.

easwars commented on xds/internal/server/listener_wrapper_test.go:
```diff
@@ -117,7 +117,11 @@ type fakeListener struct {
-	cne := <-fl.acceptCh
+	cne, ok := <-fl.acceptCh
+	if !ok {
+		nonTempErr := errors.New("a non-temporary error")
+		return nil, nonTempErr
```
Why not just inline the error:

```golang
	return nil, errors.New("accept channel closed")
```

lzhfromustc commented:
Thanks for the feedback! I have fixed the two places as you mentioned. 
For `t.Fatal()`, I replaced it with `t.Error()` just like other places in fault_test.go

songlh commented:
can we merge this pull request? I see the required changes are finished.
